### PR TITLE
WT-11458 Review shared variable semantics in hazard pointer logic

### PR DESCRIPTION
**Pull request by ajmorton**
Some small changes after reviewing hazard pointer's shared variable semantics. There's only one change in which I've moved a barrier closer to where it's required. Otherwise there's some small cosmetic changes ID'd during the review.

### CODE REVIEW

File: src/support/hazard.c [hunk 1]
```diff
@@ -28,22 +28,18 @@ hazard_grow(WT_SESSION_IMPL *session)
...
      */
     ohazard = session->hazard;
-    WT_PUBLISH(session->hazard, nhazard);
+    WT_PUBLISH(session->hazard, new_hazard);
 
-    /*
-     * Increase the size of the session's pointer array after swapping it into place (the session's
-     * reference must be updated before eviction can see the new size).
-     */
-    WT_PUBLISH(session->hazard_size, (uint32_t)(size * 2));
```
Comment (ajmorton) [inline src/support/hazard.c:46]: `hazard_size` is only ever used in `hazard_grow` and its parent function `__wt_hazard_set_func` which means there's no interthread communication using the variable. The PUBLISH barrier was actually forcing an ordering between the updated `session->hazard` array pointer and `session->hazard_inuse` so I've moved the barrier there instead.

File: src/support/hazard.c [hunk 2]
```diff
@@ -102,6 +98,13 @@ __wt_hazard_set_func(WT_SESSION_IMPL *session, WT_REF *ref, bool *busyp
         WT_ASSERT(session,
           session->nhazard == session->hazard_inuse &&
             session->hazard_inuse < session->hazard_size);
+        /*
+         * Our session->hazard pointer may have been updated by hazard_grow. If this is the case we
+         * need to ensure the updated pointer is visible to other threads before the increment of
+         * hazard_inuse as we use this value to walk the hazard array, and using the new
+         * hazard_inuse on the old hazard array can lead to reading past the end of the array.
+         */
+        WT_WRITE_BARRIER();
```
Comment (ajmorton) [inline src/support/hazard.c:107]: A `WRITE_BARRIER` seemed cleaner than a `PUBLISH` here as the `PUBLISH` ends up looking like
```
WT_PUBLISH(session->hazard_inuse, session->hazard_inuse + 1);
 hp = &session->hazard[session->hazard_inuse];
```
Open to discussion on this one.

File: src/support/hazard.c [hunk 3]
```diff
@@ -19,7 +19,7 @@ static void __hazard_dump(WT_SESSION_IMPL *);
 static int
 hazard_grow(WT_SESSION_IMPL *session)
 {
-    WT_HAZARD *nhazard;
```
Comment (ajmorton) [inline src/support/hazard.c:22]: This local variable (new hazard array pointer) conflicts with the name of the session field (number of active hazard pointers). Renaming to make it clearer

File: src/support/hazard.c [hunk 4]
```diff
@@ -28,22 +28,18 @@ hazard_grow(WT_SESSION_IMPL *session)
...
-    memcpy(nhazard, session->hazard, size * sizeof(WT_HAZARD));
+    WT_RET(__wt_calloc_def(session, size * 2, &new_hazard));
+    memcpy(new_hazard, session->hazard, size * sizeof(WT_HAZARD));
 
     /*
      * Swap the new hazard pointer array into place after initialization is complete (initialization
      * must complete before eviction can see the new hazard pointer array), then schedule the
      * original to be freed.
      */
     ohazard = session->hazard;
```
Comment (luke-pearson) [inline src/support/hazard.c:39]: Can we rename to `old_hazard`?

File: src/support/hazard.c (hunk 1 above)
Comment (luke-pearson) [reply src/support/hazard.c:46]: I agree with you on the hazard_size being singly threaded.

File: src/support/hazard.c (hunk 2 above)
Comment (luke-pearson) [reply src/support/hazard.c:107]: Wouldn't this be:
```
hp = &session->hazard[session->hazard_inuse];
WT_PUBLISH(session->hazard_inuse, session->hazard_inuse + 1);
```
It's a post increment originally.

Comment (ajmorton) [reply src/support/hazard.c:107]: Yes is would. My focus was more on the 3 separate uses of `session->hazard` if we use `WT_PUBLISH` over a `WT_WRITE_BARRIER`

File: src/support/hazard.c (hunk 4 above)
Comment (ajmorton) [reply src/support/hazard.c]: Updated in https://github.com/wiredtiger/wiredtiger/pull/9530/commits/5a7969b51aaf7cdd292a92be07e3774d13c7fe6f

File: src/support/hazard.c (hunk 2 above)
Comment (luke-pearson) [reply src/support/hazard.c:107]: So is this the `Ensure Invariant` usage pattern? I am struggling to see the exact ordering.

Comment (luke-pearson) [reply src/support/hazard.c:107]: I think I get it after reading the usage patterns doc, we need the increment to happen after the grow to guarantee a reader doesn't get an array index OOB error.

Comment (ajmorton) [reply src/support/hazard.c:107]: Yep. On the writer thread's side `session_inuse` will always be in bounds for the new array, so we just need to ensure that the new array is visible before an increment of `session_inuse`. This only applies in the case we need to increment the array.

Comment (luke-pearson) [reply src/support/hazard.c:107]: Wouldn't it be better placed in the grow function then? As that function has the context as to why we need that ordering.

Comment (ajmorton) [reply src/support/hazard.c:107]: I'm open to it. 
By doing that we lose the context of exactly which store is being ordered; When I first looked at this code I started with the ordered reads in `hazard_get_reference` and was confused by the lack of write barriers associated with the `hazard_inuse++`. I think our struggle is the logic needing ordering exists across both `hazard_set_func` and its subfunction `hazard_grow`.

I'll stare at this for a bit.

Comment (luke-pearson) [reply src/support/hazard.c:107]: I would put it in grow with a comment indicating which var it is ordering against, and a comment in hazard set explaining that there is ordering that it depends on.

Comment (ajmorton) [reply src/support/hazard.c]: Updated in https://github.com/wiredtiger/wiredtiger/pull/9530/commits/c839f94a6c8183633b1749ef93ab8b95900aeca5

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Review by clarissecheah — APPROVED** [review]
LGTM - live reviewed with Andrew. Thanks!

**Comment by ajmorton** [comment]
evergreen merge

**Pull request merged**